### PR TITLE
Merge from cuda11 to develop

### PR DESCRIPTION
**Pull request by ddhhss**

### CODE REVIEW

File: examples/01a_Si_diamond_pw_scf/INPUT [hunk 1]
```diff
@@ -3,6 +3,7 @@ INPUT_PARAMETERS
 pseudo_dir		./	
 ntype			1	
 nbands			4
+#npool                   8
```
Comment (dyzheng) [inline examples/01a_Si_diamond_pw_scf/INPUT]: This looks like a useless update

**Review by dyzheng — CHANGES_REQUESTED** [review]
Please explain the importance about the update of file  "examples/01a_Si_diamond_pw_scf/results.csv"

File: source/src_lcao/run_md_lcao.cpp [hunk 2]
```diff
@@ -40,7 +40,7 @@ void Run_MD_LCAO::opt_cell(void)
     // npwx may change according to cell change
     // this function belongs to cell LOOP
     GlobalC::wf.allocate_ekb_wg(GlobalC::kv.nks);
-
+
```
Comment (dyzheng) [inline source/src_lcao/run_md_lcao.cpp]: This looks like a useless update

File: source/src_pw/CMakeLists.txt [hunk 3]
```diff
@@ -12,7 +12,7 @@ list(APPEND objects
     diago_cg.cpp
```
Comment (dyzheng) [inline source/src_pw/CMakeLists.txt:12]: This line should be deleted either?

File: source/src_pw/diago_cg.cuh [hunk 4]
```diff
@@ -0,0 +1,178 @@
...
+class Diago_CG_CUDA
+{
+public:
+
+    Diago_CG_CUDA();
+    ~Diago_CG_CUDA();
+
+    static int moved;
+
+    float ddot_real(
```
Comment (dyzheng) [inline source/src_pw/diago_cg.cuh:20]: Please add some annotation for functions in this file, at least you can use one sentance to explain what does this function do.

File: source/src_pw/diago_cg_hip.h [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    float ddot_real(
+        const int & dim,
+        const hipblasComplex* psi_L,
+        const hipblasComplex* psi_R,
+        const bool reduce = true) ;
+
+    double ddot_real(
+        const int & dim,
+        const hipblasDoubleComplex* psi_L,
+        const hipblasDoubleComplex* psi_R,
```
Comment (dyzheng) [inline source/src_pw/diago_cg_hip.h:31]: Please add some annotation for functions in this file.

File: source/src_pw/CMakeLists.txt [hunk 6]
```diff
@@ -12,7 +12,7 @@ list(APPEND objects
...
     diago_david.cpp
     efield.cpp
-    electrons.cpp
+    # electrons.cpp
```
Comment (caic99) [inline source/src_pw/CMakeLists.txt]: What's the aim here?

File: source/src_pw/diago_david.cu [hunk 7]
```diff
@@ -0,0 +1,659 @@
...
+// optimize diag_zhegvx().
+
+// 09-05-09 wangjp
+// fixed a bug in diag_zhegvx().
+// modify the dimension of h and s as (n,n) and copy the leading N*N
+// part of hc & sc into h & s
+
+// 09-05-10 wangjp
+// As the complexmatrixs will be copied again in the subroutine ZHEGVX(...  ),
+// i.e ZHEGVX(...) will not destroy the input complexmatrixs,
```
Comment (dyzheng) [inline source/src_pw/diago_david.cu]: Please modify annotation in this file.

File: source/src_pw/diago_david.cuh [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        double2* spsi);
+
+    static void diag_zhegvx(
+        const int& n,
+        const int& m,
+        const ModuleBase::ComplexMatrix &hc,
+        const ModuleBase::ComplexMatrix &sc,
+        const int& ldh,
+        double* e,
+        ModuleBase::ComplexMatrix &vc);
```
Comment (dyzheng) [inline source/src_pw/diago_david.cuh:30]: Please add some annotation for functions in this file.

File: source/src_pw/diago_cg.cu [hunk 9]
```diff
@@ -262,66 +308,129 @@ void Diago_CG_GPU<T, T2>::diag
...
+    ONE.y = ZERO.x = ZERO.y = 0.0;
+    ONE.x = 1.0;
+    NEG_ONE.x = -1.0;
+    CHECK_CUBLAS(cublasCgemv(diag_handle, trans1, dim, m, &ONE, eigenfunction, dmx, sg, inc, &ZERO, lagrange, inc));
+    /*for (int i=0; i<m; i++)
+    {
+        lagrange[i] = ZERO;
+        for (int j=0; j<dim; j++)
+        {
+            lagrange[i] += conj( eigenfunction(i,j) ) * sg[j];
```
Comment (caic99) [inline source/src_pw/diago_cg.cu]: You may remove redundant comments.

File: source/src_pw/global.h [hunk 10]
```diff
@@ -26,13 +26,253 @@
 #include "../src_parallel/ft.h"
 
 #ifdef __CUDA
+static const char *_cublasGetErrorEnum(cublasStatus_t error)
```
Comment (caic99) [inline source/src_pw/global.h]: `...GetErrorString` may be more appropriate.

File: source/src_pw/global.h [hunk 11]
```diff
@@ -26,13 +26,253 @@
...
+{
```
Comment (dyzheng) [inline source/src_pw/global.h:32]: Are these functions globalfunctions? You would better add a namespace for them.

File: source/src_pw/hamilt.h [hunk 12]
```diff
@@ -47,8 +55,49 @@ class Hamilt
...
+#ifdef __CUDA
+	cusolverDnHandle_t cusolver_handle;
+	void diagH_subspace_cuda(
+		const int ik,
+		const int nstart,
+		const int n_band,
+		const double2* psi,
+		double2* evc,
+		double *en,
+		double2 *d_ekb_c);
```
Comment (dyzheng) [inline source/src_pw/hamilt.h:69]: Please add some annotation for functions in this file.

File: source/src_pw/hamilt_cu.bak [hunk 13]
```diff
@@ -2,7 +2,7 @@
 #include "global.h"
 #include "hamilt.h"
 #include "diago_cg.h"
-#include "diago_cg_gpu.h" // Diago_CG_GPU
+#include "diago_cg.h" // Diago_CG_CUDA
```
Comment (dyzheng) [inline source/src_pw/hamilt_cu.bak]: What is the usage of this file: hamilt_cu.bak ?

File: source/src_pw/hamilt_hip.h [hunk 14]
```diff
@@ -0,0 +1,82 @@
...
+	public:
+
+    Hamilt();
+    ~Hamilt();
+
+    void init_before_ions(void);
+
+    void init_before_electrons(void);
+
+    void clear_after_ions(void);
```
Comment (dyzheng) [inline source/src_pw/hamilt_hip.h]: Please add some annotation for functions in this file.

File: source/src_pw/hamilt_pw.cuh [hunk 15]
```diff
@@ -49,28 +55,57 @@ class Hamilt_PW
...
+        double2 *evc, // matrix
+        double *en,
+        double2 *vkb_c);
+
+    void h_1psi_cuda(
+        const int npw,
+        const float2 *psi1d,
+        float2 *hpsi,
+        float2 *spsi,
+        float2 *vkb_c);
```
Comment (dyzheng) [inline source/src_pw/hamilt_pw.cuh]: Please add some annotation for functions in this file.

File: source/src_pw/hamilt_pw.h [hunk 16]
```diff
@@ -32,7 +32,7 @@ class Hamilt_PW
 
 	friend class Diago_David;
 	friend class Diago_CG;
-    // friend class Diago_CG_GPU;
+    // friend class Diago_CG_CUDA;
```
Comment (dyzheng) [inline source/src_pw/hamilt_pw.h]: This update looks like useless.

File: CMakeLists.txt [hunk 17]
```diff
@@ -91,6 +94,35 @@ if(USE_CUDA)
   add_compile_definitions(__CUDA)
 endif()
 
+if(USE_ROCM)
+  set(CMAKE_CXX_STANDARD 11) 
+  SET(CMAKE_HIP_STANDARD 11)
```
Comment (caic99) [inline CMakeLists.txt:101]: Please note that CMake add support to `HIP` in version `3.21`. This is rather a new version.

File: source/src_pw/hamilt_pw_hip.h [hunk 18]
```diff
@@ -0,0 +1,175 @@
...
+                  ModuleBase::ComplexMatrix &evc,
+                  double *en);
+    
+    void diagH_subspace_cuda(
+        const int ik,
+        const int nstart,
+        const int n_band,
+        const hipblasDoubleComplex *psi_c, // matrix
+        hipblasDoubleComplex *evc, // matrix
+        double *en,
```
Comment (dyzheng) [inline source/src_pw/hamilt_pw_hip.h]: Please add some annotation for functions in this file.
And be careful for the code indentation format.

File: CMakeLists.txt [hunk 19]
```diff
@@ -91,6 +94,35 @@ if(USE_CUDA)
...
+  # set(CMAKE_CXX_STANDARD_REQUIRED True) 
+  list(APPEND CMAKE_PREFIX_PATH /opt/rocm/hip /opt/rocm)
```
Comment (caic99) [inline CMakeLists.txt:103]: Does it offers a argument for customize ROCm dir?

File: CMakeLists.txt [hunk 20]
```diff
@@ -91,6 +94,35 @@ if(USE_CUDA)
...
+  list(APPEND CMAKE_SYSTEM_PREFIX_PATH "/opt/rocm/hip")
+  include_directories(
+    /opt/rocm/hip/include
```
Comment (caic99) [inline CMakeLists.txt:111]: These include directories should be found after `find_package` function.

File: CMakeLists.txt [hunk 21]
```diff
@@ -91,6 +94,35 @@ if(USE_CUDA)
...
+#   find_package(rocsolver REQUIRED)
+  target_link_libraries(${ABACUS_BIN_NAME}
+    hip::device
+    hip::host
+    hip::hipfft
+    roc::hipblas
+    # roc::rocsolver
+  )
+  add_compile_definitions(__ROCM)
+  # set (HIP_HIPCC_FLAGS -hc; -fno-gpu-rdc; --amdgpu-target=gfx906)
```
Comment (caic99) [inline CMakeLists.txt]: What are these commented code for? I remember they're of some use.

File: source/input_conv.cpp [hunk 22]
```diff
@@ -47,7 +47,17 @@ void Input_Conv::Convert(void)
 	GlobalC::wf.seed = INPUT.seed;
 	GlobalC::pw.seed = INPUT.seed;
     GlobalV::NBANDS_ISTATE = INPUT.nbands_istate;
+#if ((defined __CUDA) || (defined __ROCM))
+	int temp_nproc;
+	MPI_Comm_size(MPI_COMM_WORLD, &temp_nproc);
+	if(temp_nproc != INPUT.npool)
+	{
+		std::cout << "None npool set in INPUT file, auto set npool = " << temp_nproc << std::endl;
```
Comment (caic99) [inline source/input_conv.cpp]: Do we have an output stream for log files, rather than `std::cout`? @dyzheng

File: source/src_pw/hamilt_hip.cpp [hunk 23]
```diff
@@ -0,0 +1,678 @@
...
+            ModuleBase::WARNING_QUIT("Hamilt::diago","no implemt yet.");
+		}
+    }
+    else
+    {
+        int ntry = 0;
+        int notconv = 0;
+        do
+        {
+	   		if(GlobalV::KS_SOLVER=="cg")
```
Comment (caic99) [inline source/src_pw/hamilt_hip.cpp]: You may reformat this file.

File: source/src_pw/wf_igk.cpp [hunk 24]
```diff
@@ -13,6 +13,9 @@ WF_igk::~WF_igk()
 		std::cout << " ~WF_igk()" << std::endl;
 	}
     delete[] g2kin;
+#ifdef __CUDA
+    cudaFree(this->d_g2kin);
+#endif
 }
```
Comment (dyzheng) [inline source/src_pw/wf_igk.cpp:23]: Be careful for using of d_g2kin, there is a memory leakage for no initial malloc.
Another question is can vector in STL be used in cuda or rocm?

File: source/src_pw/hamilt_hip.cpp [hunk 25]
```diff
@@ -0,0 +1,678 @@
...
+    }
+    delete[] rwork;
+    delete[] work;
+
+	ModuleBase::timer::tick("Hamilt","diagH_LAPACK");
+    return;
+}
+
+/*
+void Hamilt::diagH_CUSOLVER(
```
Comment (caic99) [inline source/src_pw/hamilt_hip.cpp]: What's this function for?

**Review by dyzheng — CHANGES_REQUESTED** [review]

File: source/src_pw/hamilt_pw.cuh [hunk 26]
```diff
@@ -49,28 +55,57 @@ class Hamilt_PW
...
-    void h_psi_gpu(
-		const CUFFT_COMPLEX *psi,
-		CUFFT_COMPLEX *hpsi,
-		const int m = 1); // qianrui add a default parameter 2021-3-31
-
-    void s_1psi_gpu(
+    void h_psi_cuda(
+		const float2 *psi,
+		float2 *hpsi,
+        float2 *vkb_c,
```
Comment (caic99) [inline source/src_pw/hamilt_pw.cuh]: Please using tab but not whitespace for indentation. Reformatting would be appreciated.

**Review by caic99 — CHANGES_REQUESTED** [review]

File: source/src_pw/CMakeLists.txt (hunk 3 above)
Comment (ddhhss) [reply source/src_pw/CMakeLists.txt:12]: Some functions in this file shoule be remain in GPU version, I just deleted "diago_cg.cpp" in L82.

File: source/input_conv.cpp (hunk 22 above)
Comment (dyzheng) [reply source/input_conv.cpp]: Yes, "cout" should not be used arbitrarily.

File: source/src_pw/wf_igk.cpp (hunk 24 above)
Comment (ddhhss) [reply source/src_pw/wf_igk.cpp:23]: I have allocated memory for d_g2kin in wavefunc.cpp:L70, is it reasonable to do so? wavefunc inherits from WF_igk.

Comment (ddhhss) [reply source/src_pw/wf_igk.cpp:23]: Cuda does not yet support STL on the device side..

Comment (Asuna981002) [reply source/src_pw/wf_igk.cpp:19]: Cuda does not yet support STL on the device side.

File: examples/01a_Si_diamond_pw_scf/INPUT (hunk 1 above)
Comment (ddhhss) [reply examples/01a_Si_diamond_pw_scf/INPUT]: It is indeed a useless modification, it has been fixed.

File: source/src_lcao/run_md_lcao.cpp (hunk 2 above)
Comment (ddhhss) [reply source/src_lcao/run_md_lcao.cpp]: It has been fixed.

File: source/src_pw/wf_igk.cpp (hunk 24 above)
Comment (caic99) [reply source/src_pw/wf_igk.cpp:23]: > Cuda does not yet support STL on the device side..

Well it does have STL on both device and host, see https://github.com/NVIDIA/libcudacxx .
Personally, I have no experiece on how to use it.
FYI @dyzheng

File: examples/01a_Si_diamond_pw_scf/INPUT (hunk 1 above)
Comment (caic99) [reply examples/01a_Si_diamond_pw_scf/INPUT]: > It is indeed a useless modification, it has been fixed.

@Asuna981002 You may 'Resolve conversation' by clicking the button below after patching.

File: source/input_conv.cpp (hunk 22 above)
Comment (caic99) [reply source/input_conv.cpp]: > Yes, "cout" should not be used arbitrarily.

So, which function shall we use in this case? @dyzheng

File: source/src_pw/hamilt_cu.bak (hunk 13 above)
Comment (ddhhss) [reply source/src_pw/hamilt_cu.bak]: This file has been deleted.

File: source/src_pw/hamilt_hip.cpp (hunk 25 above)
Comment (ddhhss) [reply source/src_pw/hamilt_hip.cpp]: This function is not used in the ROCm version，it and has been deleted.

File: source/src_pw/CMakeLists.txt (hunk 6 above)
Comment (ddhhss) [reply source/src_pw/CMakeLists.txt]: Electrons.cpp should be used in CPU version, I put this file in L80.

File: source/src_pw/global.h (hunk 11 above)
Comment (ddhhss) [reply source/src_pw/global.h:32]: I just added namespace CudaCheck and HipCheck to fix it.

File: CMakeLists.txt (hunk 19 above)
Comment (ddhhss) [reply CMakeLists.txt:103]: Cmake 3.21 mentioned the support for CXX=hipcc, but did not mention that there are new variable names, I think this part can be left as it is for the time being? 
[https://cmake.org/cmake/help/v3.21/release/3.21.html#new-features](url)
[https://rocmdocs.amd.com/en/latest/Installation_Guide/Using-CMake-with-AMD-ROCm.html](url)

Comment (caic99) [reply CMakeLists.txt:103]: You're right. I suggest adding this url (https://rocmdocs.amd.com/en/latest/Installation_Guide/Using-CMake-with-AMD-ROCm.html) as comment for reference, do you think it's OK?

File: CMakeLists.txt (hunk 20 above)
Comment (caic99) [reply CMakeLists.txt:111]: i. Is it OK to leave out the `include_directories` lines? From my knowledge, `target_link_libraries hip_<something>` also performs such operation on CMake 3.18+ .
ii. Is there any generated variable such as `${HIP_INCLUDE_DIR}` or something just like that(I'm not sure about the name) for you to use in `include_directories`?

File: source/src_pw/hamilt.h [hunk 27]
```diff
@@ -56,7 +56,9 @@ class Hamilt
 		ModuleBase::ComplexMatrix &hvec); // output: eigenvectors
 
 #ifdef __CUDA
-	cusolverDnHandle_t cusolver_handle;
+	cusolverDnHandle_t cusolver_handle; // cusolver handle
```
Comment (caic99) [inline source/src_pw/hamilt.h:59]: The comment here seems to be a little silly.

**Review by caic99 — APPROVED** [review]

**Review by dyzheng — APPROVED** [review]

**Pull request merged**